src/signatures/secp256k1_ecdsa.rs

```diff
@@ -0,0 +1,210 @@
+                to: "secp256k1 ecdsa secret key",
+            })
+            .map(Self)
+    }
+
+    // Signature operation can fail for some reason.
+    // Not sure what to do. Panic? Retry?
+    pub fn try_sign(&self, msg: &[u8]) -> crate::Result<Signature> {
+        self.0
+            .sign_recoverable(msg)
+            .map_err(|_| crate::Error::SignatureError { alg: "secp256k1 ecdsa" })
+            .map(|(sig, rid)| Signature(sig, rid))
+    }
+
+    pub fn sign(&self, msg: &[u8]) -> Signature {
+        self.try_sign(msg).expect("secp256k1 ecdsa sign failed")
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct PublicKey(k256::ecdsa::VerifyingKey);
+
+impl PublicKey {
+    pub const LENGTH: usize = 33;
+
+    pub fn verify(&self, sig: &Signature, msg: &[u8]) -> bool {
+        use k256::ecdsa::signature::Verifier;
+        self.0.verify(msg, &sig.0).is_ok()
+    }
+
+    pub fn to_bytes(self) -> [u8; PublicKey::LENGTH] {
+        let encoded_point = self.0.to_encoded_point(true);
+        let slice = encoded_point.as_bytes();
+        let mut bytes = [0_u8; PublicKey::LENGTH];
+        bytes.copy_from_slice(slice);
+        bytes
+    }
+
+    pub fn try_from_bytes(bytes: &[u8; PublicKey::LENGTH]) -> crate::Result<Self> {
+        k256::ecdsa::VerifyingKey::from_sec1_bytes(bytes)
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "compressed bytes",
+                to: "Secp256k1 public key",
+            })
+    }
+
+    pub fn try_from_slice(bytes: &[u8]) -> crate::Result<Self> {
+        k256::ecdsa::VerifyingKey::from_sec1_bytes(bytes)
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "compressed slice",
+                to: "Secp256k1 public key",
+            })
+    }
+
+    // credit: [secret_key_to_address](https://github.com/gakonst/ethers-rs/)
+    pub fn to_address(&self) -> Address {
+        // let public_key = secret_key.verifying_key();
+        let public_key = self.0.to_encoded_point(/* compress = */ false);
+        let public_key = public_key.as_bytes();
+        debug_assert_eq!(public_key[0], 0x04);
+        // let hash = keccak256(&public_key[1..]);
+        use tiny_keccak::{Hasher, Keccak};
+        let mut keccak = Keccak::v256();
+        keccak.update(&public_key[1..]);
+        let mut hash = [0u8; 32];
+        keccak.finalize(&mut hash);
+
+        let mut bytes = [0u8; 20];
+        bytes.copy_from_slice(&hash[12..]);
+        Address::from(bytes)
+    }
+}
+
+impl TryFrom<[u8; PublicKey::LENGTH]> for PublicKey {
+    type Error = crate::Error;
+    fn try_from(bytes: [u8; PublicKey::LENGTH]) -> crate::Result<Self> {
+        Self::try_from_bytes(&bytes)
+    }
+}
+
+impl Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let encoded_point = self.0.to_encoded_point(true);
+        encoded_point.as_bytes().hash(state);
```

Comment:
For my own curiosity, if you could explain the different between encoded point and sec1 bytes please 🙏🏻

Reply:
Encoded point is the SEC1 encoded point, it can be in compressed, uncompressed or compact form. I'm not sure about compact form, looks like it loses information about the y coordinate. Compressed SEC1 point seems to do the trick, it is [tag + x-coordinate] where tag=2 encodes even y coord, tag=3 encodes odd y coord. Compressed SEC1 form is used in Ethereum for public keys and in SLIP-10 in non-hardened key derivation.

Reply:
I think you meant why I didn't use `to_sec1_bytes()` here and in `PublicKey::to_bytes()`. The reason is to avoid using `Box`, internally `to_sec1_bytes()` it constructs an encoded point with a variable length (because its tag/form can vary). To avoid using `Box` and allocating unnecessary memory I create encoded point directly.